pkg/webhook/resources/upgrade/validator.go

```diff
@@ -307,6 +311,38 @@ func (v *upgradeValidator) checkMachines() error {
 	return nil
 }
 
+func (v *upgradeValidator) checkSingleReplicaVolumes() error {
+	// Upgrade should be rejected if any single-replica volume exists
+	nodes, err := v.nodes.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	// Skip for single-node cluster
+	if len(nodes) == 1 {
+		return nil
+	}
+
+	var singleReplicaVolumeNames []string
+
+	volumes, err := v.lhVolumes.List(util.LonghornSystemNamespaceName, labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, volume := range volumes {
+		if volume.Spec.NumberOfReplicas == 1 {
+			singleReplicaVolumeNames = append(singleReplicaVolumeNames, volume.Namespace+"/"+volume.Name)
+		}
+	}
+
+	if len(singleReplicaVolumeNames) > 0 {
+		return werror.NewInternalError(fmt.Sprintf("single-relica volumes (%s) running, please shutdown the correspondingn workload before upgrade", strings.Join(singleReplicaVolumeNames, ", ")))
```

Comment:
```suggestion
		return werror.NewInternalError(fmt.Sprintf("single-replica volumes (%s) running, please shutdown the corresponding workload before upgrade", strings.Join(singleReplicaVolumeNames, ", ")))
```

Reply:
Instead of displaying Longhorn volumes here, I suggest we display pvc of volumes. It shows the correct namespace and volume name to users. Also, we can ignore detached volumes.

Reply:
Updated, thank you.

Reply:
Updated, thank you!